### PR TITLE
Added add/replace credential validation on jt launch and schedule

### PR DESCRIPTION
**Pull request by mabashian**
##### SUMMARY
Most of the work in this PR stems from the need to bring the UI in line with the API's decision to only allow users to add/replace default credentials when launching a job, scheduling a job, or creating a workflow job template node (as long as the JT in question has ask_credential_on_launch set to true).  This means that if a JT has a default machine credential with ask_credential_on_launch set to true and they go to launch that JT they won't be able to explicitly remove that default machine credential without replacing it with another machine credential.  Additional credentials can be added as the user sees fit.  As a result, in the UI we wanted to avoid situations where users removed credentials, attempted to launch and got a 400 error.  To achieve this, we prevent the user from moving past the credentials tab until they've replaced any removed default credentials or revert them back to the default.  Note that this applies to all credential types including vault credentials where the user can select up to N vault credentials as long as they all have different vault_id's. 

This PR addresses the following:

1) When launching a job template from the templates list, the user will now be forced to replace any default credentials that have been removed before launching.
2) When creating a JT schedule, the user will now be forced to replace any default credentials that have been removed before saving.
3) Fixed a bug with the multi-select lists where clicking the row and clicking the checkbox called two different functions.

When a user removes a default credential in the prompt modal they'll see a message like this and the save button disabled until it's been addressed:

screen shot 2018-03-02 at 10 37 30 am

related https://github.com/ansible/awx/issues/169

Ping @tvo318 - We'll definitely want to document this behavior as it diverges from previous functionality and some users will likely have questions.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - UI

##### AWX VERSION
awx: 1.0.4.38

### CODE REVIEW

**Comment by tvo318** [comment]
Consider this noted, @mabashian. Doc issue #[337](https://github.com/ansible/product-docs/issues/337) created.

File: awx/ui/client/src/scheduler/factories/schedule-post.factory.js [hunk 1]
```diff
@@ -64,43 +65,65 @@ export default
...
+                                // in the template defaults
+                                let credentialsToPost = promptData.prompts.credentials.value.filter(function(credFromPrompt) {
+                                    let defaultCreds = promptData.launchConf.defaults.credentials ? promptData.launchConf.defaults.credentials : [];
+                                    return !defaultCreds.some(function(defaultCred) {
+                                        return credFromPrompt.id === defaultCred.id;
+                                    });
+                                });
+
+                                let promises = [];
+                                let schedule = new Schedule();
```
Comment (marshmalien) [inline awx/ui/client/src/scheduler/factories/schedule-post.factory.js]: Is schedule reassigned at some point? If not, we could reconsider using `let`. This also applies to any identifier that is not reassigned in the pr.

File: awx/ui/client/src/scheduler/schedulerEdit.controller.js [hunk 2]
```diff
@@ -283,7 +284,23 @@ function($filter, $state, $stateParams, Wait, $scope, moment,
...
                         });
 
-                        prompts.credentials.value = scheduleCredentials.results.length > 0 ? scheduleCredentials.results : prompts.credentials.value;
+                        let defaultCredsWithoutOverrides = [];
+
+                        prompts.credentials.value.forEach((defaultCred) => {
+                            let typeMatches = false;
+                            scheduleCredentials.forEach((scheduleCred) => {
+                                if(defaultCred.credential_type === scheduleCred.credential_type) {
+                                    if((!defaultCred.vault_id && !scheduleCred.inputs.vault_id) || (defaultCred.vault_id && scheduleCred.inputs.vault_id && defaultCred.vault_id === scheduleCred.inputs.vault_id)) {
```
Comment (marshmalien) [inline awx/ui/client/src/scheduler/schedulerEdit.controller.js]: Here, there are a lot of conditions that are hard to follow. What do you think about breaking this out into another method?

File: awx/ui/client/src/templates/prompt/steps/credential/prompt-credential.controller.js [hunk 3]
```diff
@@ -88,20 +88,57 @@ export default
...
                             wipePasswords(scope.promptData.prompts.credentials.value[i]);
                             scope.promptData.prompts.credentials.value.splice(i, 1);
                         }
                     }
 
-                    scope.promptData.prompts.credentials.value.push(_.cloneDeep(selectedRow));
+                    scope.promptData.prompts.credentials.value.push(selectedCred);
                     updateNeededPasswords(selectedRow);
+
+                    for (let i = scope.promptData.credentialTypeMissing.length - 1; i >= 0; i--) {
```
Comment (marshmalien) [inline awx/ui/client/src/templates/prompt/steps/credential/prompt-credential.controller.js:136]: This could probably be simplified with lodash's `_.remove` method.

File: awx/ui/client/src/shared/generator-helpers.js [hunk 4]
```diff
@@ -560,7 +560,13 @@ angular.module('GeneratorHelpers', [systemStatus.name])
                 }
                 html += "\" ";
                 html += field.columnNgClass ? " ng-class=\"" + field.columnNgClass + "\"": "";
-                html += (options.mode === 'lookup' || options.mode === 'select') ? " ng-click=\"toggle_row(" + list.iterator + ")\"" : "";
+                if(options.mode === 'lookup' || options.mode === 'select') {
```
Comment (marshmalien) [inline awx/ui/client/src/shared/generator-helpers.js:568]: Thanks for fixing this bug 🎉

**Review by marshmalien — APPROVED** [review]

File: awx/ui/client/src/scheduler/factories/schedule-post.factory.js (hunk 1 above)
Comment (mabashian) [reply awx/ui/client/src/scheduler/factories/schedule-post.factory.js]: Ah good catch - i'll run back through and take a look at where we might be able to do that.

**Comment by jaredevantabor** [comment]
I'm seeing an error message 2x for each credential that is removed. Code looks good to me tho +1
before: 
image

after: 
image

**Comment by jaredevantabor** [comment]
I'm also unable to choose two vault creds at the same time, the multi-select checkboxes are essentially acting like radio checkboxes...
image
